File: arrays-and-hashing/49-group-anagrams/main.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>
using namespace std;

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> anagrams;
        
        for (const string& s : strs) {
            string sorted_s = s;  // Make a copy of the string
            sort(sorted_s.begin(), sorted_s.end());  // Sort the string
            
            anagrams[sorted_s].push_back(s);  // Group anagrams together
        }
        
        vector<vector<string>> result;
        for (auto& entry : anagrams) {
            result.push_back(entry.second);  // Collect all the anagram groups
        }
        
        return result;
    }
};



int main() {
    Solution solution;
    vector<string> strs = {"eat", "tea", "tan", "ate", "nat", "bat"};
    
    vector<vector<string>> result = solution.groupAnagrams(strs);
    
    for (const auto& group : result) {
        for (const auto& word : group) {
            cout << word << " ";
        }
        cout << endl;
    }
    
    return 0;
}

File: arrays-and-hashing/49-group-anagrams/main.py
class Solution(object):
    def groupAnagrams(self, strs):
        """
        :type strs: List[str]
        :rtype: List[List[str]]
        """
        anagrams = {}
        for i in strs:
            key = ''.join(sorted(i))

            if key in anagrams:
                anagrams[key].append(i)
            else:
                anagrams[key] = [i]
        return list(anagrams.values())


solution = Solution()
strs = ["eat", "tea", "tan", "ate", "nat", "bat"]
print(solution.groupAnagrams(strs))

# Steps:
# 1 - Initialize an empty dictionary to store the anagram groups.
# 2 - For each string:
#         - Compute its signature (either by sorting the string or using a character frequency count).
#         - Check if the signature is already in the dictionary:
#             - If it is, append the current string to the list associated with that signature.
#             - If not, create a new list and add the string to it.
# 3 - After processing all strings, the dictionary will contain all the groups of anagrams.